### ISSUE TITLE
Why sometimes there are multiple  bitnami_magento databases get created in 1 dock?

### ISSUE THREAD

**Issue by SamuelUnsw** [bitnami/bitnami-docker-mariadb#117]
**Description**

**Steps to reproduce the issue:**

1. Use docker-compose.yml to create MariaDB and corresponding Magento docks

**Describe the results you received:**

Sometimes there will be multiple bitnami_magento databases get created. For example:

bitnami_magento
bitnami_magento_4aj04
bitnami_magento_QvQCR
bitnami_magento_SPbBw

**Describe the results you expected:**

However sometimes there is only 1 -- bitnami_magento get created, which I expect should be the right case.

**Additional information you deem important (e.g. issue happens only occasionally):**

Could someone please explain to me why and how could I fix it if it's not a bug?

Thanks a lot!

**Comment by dgonzalezruiz** [comment]
The reason why a random suffix is added is that the magento container detects there is an already existing `bitnami_magento` database. This is probably related to the way you are launching the container. Are you using the `docker-compose.yml` file found in the magento repo? 

As you say, only one database will be created if the containers are launched as expected.

We’ll be looking forward to receiving your reply.

Best regards,

David González

**Comment by SamuelUnsw** [comment]
Hi David, the replication steps I wrote in another post: https://github.com/bitnami/bitnami-docker-magento/issues/61

In short:

1, after using "docker-compose up" there are 2 containers get created. One is Magento and another is MariaDB. The multiple 'bitnami_magento' database(some with suffix) are found in MariaDB docker. And only 1 of them contains table/data which I guess is for LUMA website deployed on Magento docker to use. ( rest DB seems to be empty.)

2, I am not 100% sure that is: if the port I set in docker-compose.yml file is to use default '80', then it seems only 1 database will be created. But if using other publish port like '82:80' then multiple database may get created.


3, The purpose of running multiple yml files with different ports is because we want multiple LUMA websites setup for multiple users and they could access their own by using different port number. But it comes with such strange issue and the one I raised in the link above in 'Magento' section.

**Comment by SamuelUnsw** [comment]
Hi David,

This morning I found 1 thing: if I run these command every time before running a yml file (with different port number inside), there "seems" no duplication will be created:

docker volume rm xyz_mariadb_data
docker volume rm xyz_apache_data
docker volume rm xyz_magento_data
docker volume rm xyz_php_data

---------------------------------------------------------------------------

Is it because if I run the docker-compose before, the volume data will be kept on my local machine even the container get deleted? (which cause MariaDB creates multiple database for bitnami_magento?)

Please correct me if my guess is wrong.

**Comment by dgonzalezruiz** [comment]
Your assumption is correct. Docker volumes are only detached from the container when the container is deleted, but not deleted. Therefore, a new container would just reattach that volume to itself, instead of create a new one. To avoid this, you have to delete them, as you did.

To take the volumes related to a docker-compose file down, together with those volumes, you must execute:

```
docker-compose down -v
```

When those already used volumes are reattached to your container, there's logic inside them to try to reuse them, without modifying any bit of the data found inside. That would be enough to save the state of MariaDB, but in your case it's leading to undesired behaviour.

May I close this issue, or is there anything we can help you with?

Cheers!

**Comment by SamuelUnsw** [comment]
Hi David,

The command you gave me
`docker-compose down -v`

I think is to remove/delete all current containers and their volumes. Can I pass certain container name/ID to this command?

BTW, David, are you looking after Bitnami--MariaDB section only? so some other developer takes care of Bitnami--Magento section for example?


Yes, you could close this thread after this :-)  Thanks again for your help!

**Comment by sameersbn** [comment]
Hi @SamuelUnsw, like @dgonzalezruiz mentioned the issue you see is due to the way you are running the containers. You should be mounting persistent volumes for all the containers that are involved in the deployment and I have a feeling that you are having a long running mariadb instance and the magento instance is being created and destroyed, which would explain why the magento database name is suffixed with the random string.

Perhaps you should use the docker-compose.yml sample from the https://github.com/bitnami/bitnami-docker-magento:

```yaml
version: '2'
services:
  mariadb:
    image: 'bitnami/mariadb:latest'
    environment:
      - ALLOW_EMPTY_PASSWORD=yes
    volumes:
      - 'mariadb_data:/bitnami/mariadb'
  magento:
    image: 'bitnami/magento:latest'
    ports:
      - '10080:80'
      - '10443:443'
    volumes:
      - 'magento_data:/bitnami/magento'
      - 'apache_data:/bitnami/apache'
      - 'php_data:/bitnami/php'
    depends_on:
      - mariadb
    environment:
      - MAGENTO_HOST=app.damagehead.com
volumes:
  mariadb_data:
    driver: local
  magento_data:
    driver: local
  apache_data:
    driver: local
  php_data:
    driver: local
```

The above sample will create ephemeral volumes which will be removed with `docker-compose down -v`. Instead of using ephemeral volume mount locations you should specify the host path for these volumes like so:

```yaml
version: '2'
services:
  mariadb:
    image: 'bitnami/mariadb:latest'
    environment:
      - ALLOW_EMPTY_PASSWORD=yes
    volumes:
      - './magento/mariadb:/bitnami/mariadb'
  magento:
    image: 'bitnami/magento:latest'
    ports:
      - '10080:80'
      - '10443:443'
    volumes:
      - './magento/magento:/bitnami/magento'
      - './magento/apache:/bitnami/apache'
      - './magento/php:/bitnami/php'
    depends_on:
      - mariadb
    environment:
      - MAGENTO_HOST=app.damagehead.com
```

Please let us know if this helps.

**Comment by carrodher** [comment]
We are glad that you were able to fix the issue. If you have any other questions, please do not hesitate to let us know.

**Comment by SamuelUnsw** [comment]
Hi all masters,

Actually what I want to archive is: 

1) Create a Mariadb container from latest Mariadb image; (with Magento of course.)

2) Make changes to the bitnami_magento database in current Mariadb container; --- no matter volume is ephemeral or persistent. (but you could correct me.). I *assume* all changes are made & saved to that volume & image.


**3) Make a "copy" of current image and assume all changes I made are in the new image. So the new container created from that new image (together with its new volume) ideally should also contain all changes I made in that original container/image.


Is it possible?


Thanks a lot again!

**Comment by sameersbn** [comment]
> Is it possible?

Short answer: No this is not possible.

Long answer:  The MariaDB Dockerfile [mounts a volume](https://github.com/bitnami/bitnami-docker-mariadb/blob/master/10/Dockerfile#L32) for a data and configuration persistence at `/bitnami/mariadb`. This means the database changes are stored in the volume.  Saving the image will not save these changes in the docker image as they are present in the external volume.

Instead of copying or saving the docker image, you can copy the volume to have the same effect. But make sure that you preserve the ownership and permissions of the files in the copy.

**Comment by SamuelUnsw** [comment]
Hi sameersbn, 

For "copy volume" you mentioned here, actually after run docker volume inspect command, I got:

 "Mountpoint": "/var/lib/docker/volumes/base_mariadb_data/_data",


Does that mean I just need to copy those data and create like this for my new container--test1?

"/var/lib/docker/volumes/test1_mariadb_data/_data",



If so, how come it says "Access Denied" if I want to access folder to level "/var/lib/docker/volumes"? And also if I successfully make the copy, how should I write my docker-compose.yml file for the "volume" part?



Another thinking is: maybe for Magento_data, apach_data and php_data I could use volume copy. However for MariaDb, could I just use "mysql backup" first and then restore to the new container? (Of course I still need to update core_config_data record for a different baseUrl value.)



Thank you so much again!!

**Comment by sameersbn** [comment]
Copy the contents of the directory while preserving the file and directory permissions, using `rsync`:

```bash
mkdir my-mariadb-data
sudo rsync -av /var/lib/docker/volumes/redmine_mariadb_data/_data/ ./my-mariadb-data/
```

Now you will have the mariadb data volume data in the `my-mariadb-data` directory.

> Does that mean I just need to copy those data and create like this for my new container--test1?

No sir. You need to specify this volume while starting the container, f.e. in the docker-compose.yml file:

```yaml
...
services:
  mariadb:
    image: 'bitnami/mariadb:latest'
    environment:
      - ALLOW_EMPTY_PASSWORD=yes
    volumes:
      - './my-mariadb-data:/bitnami/mariadb'
...
```

> how come it says "Access Denied" if I want to access folder to level "/var/lib/docker/volumes"? 

This is because the directory has ownership and modes set which do not allow the regular user to access these files. You need to be `root` user or use `sudo` to elevate your user privileges.

As a sidenode, consider specifying host paths for the volumes in the docker-compose.yml file, see second part of https://github.com/bitnami/bitnami-docker-mariadb/issues/117#issuecomment-290737790 for an example.

Also I would suggest you take a look at https://docs.docker.com/engine/reference/run/#volume-shared-filesystems and also https://docs.docker.com/compose/compose-file/#volumes for a basic understanding about docker volumes and how they can be worked with.

Have a good day.

**Comment by SamuelUnsw** [comment]
Dear sameersbn MASTER,

Thanks for you detailed instruction and I will try tmr when back to office,


I also found 1 script online to copy/clone volume (https://github.com/gdiepen/docker-convenience-scripts). For example it could clone my **/var/lib/docker/volumes/# base_mariadb_data/_data** to **/var/lib/docker/volumes/# test1_mariadb_data/_data** 

Then I still use the 1st version of volumes in your [#117 (comment)](https://github.com/bitnami/bitnami-docker-mariadb/issues/117#issuecomment-290737790), but run as: docker-compose **-p test1** up

The container seems still can use the new volume files I just copied, BUT will still cause the original issue I mentioned at very beginning --- that a duplicated bitnami_magento db will be created with random suffix.


Also I **assume** copy magento/php/apach volume should have no such kind of issue.

**Comment by sameersbn** [comment]
> but run as: docker-compose -p test1 up

I would not recommend messing with the contents of `/var/lib/docker`. 

> BUT will still cause the original issue I mentioned at very beginning --- that a duplicated bitnami_magento db will be created with random suffix.

Yes, you will see the issue.
You need to make sure the magento volumes are also copied to avoid the issue.

> Also I assume copy magento/php/apach volume for new containers to use should have no such kind of drama. Please correct me if I am wrong.

You are right. No drama, only happiness.

**Comment by SamuelUnsw** [comment]
Hi sameersbn master,

I am really sorry to ask you again as I am totally confused. I have followed you example above and ONLY made 1 small changes to the volume path:

    volumes:
      - './magento**_base**/mariadb:/bitnami/mariadb'

    volumes:
      - './magento**_base**/magento:/bitnami/magento'
      - './magento**_base**/apache:/bitnami/apache'
      - './magento**_base**/php:/bitnami/php'



However after running "dock-compose up":


1) docker_volume ls  ---- doesn't show any volume in the list;

2) I use docker inspect --- and it tells me:

        "Mounts": [
            {
                "Type": "bind",
                "Source": "/home/sam/docker/magento_base/mariadb",
                "Destination": "/bitnami/mariadb",
                "Mode": "rw",
                "RW": true,
                "Propagation": ""


I can understand the "Source" and "Destination" value based on volume I set in yml file.  Ans on mariadb container it does have such destination folder "/bitnami/mariadb". But how come on my host box, there is no folder created: /home/sam/docker/magento_base/mariadb? Even if I manually create these folder first, they are still empty after installation.


P.S. I also looked at the article you sent me, but could not find the answer.


Thanks again MASTER! :-)
